examples_test/config_test.go

```diff
@@ -0,0 +1,144 @@
+	"github.com/lyraproj/pcore/px"
+	"github.com/lyraproj/pcore/types"
+)
+
+// sayHello is a very simple "lookup_key" function that just returns the result of concatenating
+// the key with the string " world".
+func sayHello(pc hieraapi.ServerContext, key string) px.Value {
+	return types.WrapString(key + ` world`)
+}
+
+// TestConfig_hardwired utilizes Hiera in the simplest way possible. No configuration file and no options. Just
+// a function performs a lookup of a key.
+func TestConfig_hardwired(t *testing.T) {
+	// Use the hiera.DoWithParent to initialize a Hiera context (session) with the sayHello as the top-level function and
+	// perform a lookup.
+	//
+	// The DoWithParent is meant to be called once and the created context can then be used for any number of lookups that
+	// uses the same configuration. The sessions life-cycle can be compared to the compilers life-cycle in puppet.
+	hiera.DoWithParent(context.Background(), sayHello, nil, func(c px.Context) {
+		result := hiera.Lookup(hiera.NewInvocation(c, nil, nil), `hello`, nil, nil)
+		if result == nil || `hello world` != result.String() {
+			t.Fatalf("unexpected result %v", result)
+		}
+	})
+}
+
+// TestHelloWorld_semiHardWired uses the special "lookup_key" function MuxLookupKey. This function use the configuration
+// option LookupProvidersKey where it expects to find a slice of "lookup_key" functions to use.
+func TestConfig_semiHardWired(t *testing.T) {
+	// Create "global" options for this Hiera session (global as in "kept in the top-level context". An executable may
+	// have an unlimited set of such contexts and there are no global variables involved.
+	configOptions := make(map[string]px.Value)
+
+	// The LookupProvidersKey stores a go slice of Hiera "lookup_key" functions that serve as the top level functions.
+	configOptions[provider.LookupProvidersKey] = types.WrapRuntime([]hieraapi.LookupKey{sayHello})
+
+	// Initialize a Hiera session with the MuxLookupKey as the top-level function and perform a lookup and
+	// the created configOptions.
+	hiera.DoWithParent(context.Background(), provider.MuxLookupKey, configOptions, func(c px.Context) {
+		result := hiera.Lookup(hiera.NewInvocation(c, nil, nil), `hello`, nil, nil)
+		if result == nil || `hello world` != result.String() {
+			t.Fatalf("unexpected result %v", result)
+		}
+	})
+}
+
+/*
+ The remaining in tests in this file use the ConfigLookupKey provider. This provider will consult the configuration
+ options HieraRoot, HieraConfigFileName, and HieraConfig to determine the path of the configuration file. Use of
+ HieraConfig is mutually exclusive to use of HieraRoot and HieraConfigFileName.
+
+ HieraRoot will default to the current working directory
+
+ HieraConfigFileName will default to "hiera.yaml"
+
+ If the HieraRoot or HieraConfig are relative paths, they will be considered relative to the current directory.
+
+ The HieraConfigFileName must be relative to the HieraRoot (or the current directory).
```

Comment:
Ambiguous - must the file be a child of the HieraRoot (i.e. if it is a full path), or is this telling me that a relative path is relative to the HieraRoot ? or what?